File: bindings/python/tests/test_rlib.py
"""Tests for rlib Python bindings."""
import pytest
import rlib

def test_add():
    """Test addition function."""
    assert rlib.add(2, 3) == 5
    assert rlib.add(-1, 1) == 0
    assert rlib.add(0, 0) == 0

def test_multiply():
    """Test multiplication function."""
    assert rlib.multiply(2, 3) == 6
    assert rlib.multiply(-2, 3) == -6
    assert rlib.multiply(0, 5) == 0

def test_exponent():
    """Test exponent function."""
    assert rlib.exponent(2, 3) == 8
    assert rlib.exponent(5, 2) == 25
    assert rlib.exponent(10, 0) == 1
    assert rlib.exponent(3, 4) == 81

def test_divide():
    """Test division function."""
    assert rlib.divide(6, 2) == 3
    assert rlib.divide(10, 5) == 2
    assert rlib.divide(-10, 2) == -5

def test_divide_by_zero():
    """Test division by zero raises error."""
    with pytest.raises(ValueError, match="Division by zero"):
        rlib.divide(10, 0)

File: bindings/cpp/tests/test_rlib.cpp
#include "../rlib.hpp"
#include <cassert>
#include <iostream>
#include <stdexcept>

void test_add() {
    assert(rlib::add(2, 3) == 5);
    assert(rlib::add(-1, 1) == 0);
    assert(rlib::add(0, 0) == 0);
    std::cout << "✓ test_add" << std::endl;
}

void test_multiply() {
    assert(rlib::multiply(2, 3) == 6);
    assert(rlib::multiply(-2, 3) == -6);
    assert(rlib::multiply(0, 5) == 0);
    std::cout << "✓ test_multiply" << std::endl;
}

void test_divide() {
    assert(rlib::divide(6, 2) == 3);
    assert(rlib::divide(10, 5) == 2);
    assert(rlib::divide(-10, 2) == -5);
    std::cout << "✓ test_divide" << std::endl;
}

void test_divide_by_zero() {
    try {
        rlib::divide(10, 0);
        std::cerr << "✗ Expected exception for division by zero" << std::endl;
        std::exit(1);
    } catch (const rlib::Error& e) {
        assert(std::string(e.what()) == "Division by zero");
        std::cout << "✓ test_divide_by_zero" << std::endl;
    }
}

int main() {
    test_add();
    test_multiply();
    test_divide();
    test_divide_by_zero();
    std::cout << "\nAll tests passed!" << std::endl;
    return 0;
}
